hwtypes/smt_utils.py

```diff
@@ -0,0 +1,76 @@
+from . import SMTBit
+import operator
+from functools import partial, reduce
+import abc
+
+_or_reduce = partial(reduce, operator.or_)
+_and_reduce = partial(reduce, operator.and_)
+
+class FormulaConstructor:
```

Comment:
For abc.abstractmethod to do anything you need to inherit from `abc.ABC` or use the abc metaclass `metaclass=abc.ABCMeta` (which is what inheriting from abc.ABC does).  

See: